ibis/expr/signature.py

```diff
@@ -194,9 +194,20 @@ def __init__(self, **kwargs):
     # TODO(kszucs): split for better __init__ composability but we can
     # directly call it from __init__ too
     def __post_init__(self):
+        import ibis.expr.types as ir
+
         args = tuple(getattr(self, name) for name in self.argnames)
         object.__setattr__(self, "args", args)
         object.__setattr__(self, "_hash", hash((type(self), args)))
+        object.__setattr__(
```

Comment:
Is there a significant performance gain by caching it?
If not I'd prefer keeping it as a method, otherwise could we move it to `Node`? `Annotable` should be agnostic about `ibis.expr`.

Reply:
Additionally if we're using it only from repr which isn't performance sensitive I think it doesn't worth the tradeoff.

Reply:
Yes, there's definitely a nice perf bump (3-4x IIRC) for storing this. There are other places outside of repring that it can be used, we can look at that as a follow up.

Reply:
Moved it to `Node`.